From 0f48144c961308bd60e8b9ab45bfe8ee269b82cb Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Wed, 11 Jun 2025 16:00:21 +0200
Subject: [PATCH 0221/1010] net: fec: defer DMA unmapping to the inband stage

Calling dma_unmap_page_attrs() from the out-of-band stage is now
considered potentially harmful by Dovetail. Defer such release to the
in-band stage using an irq_work handler.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 drivers/net/ethernet/freescale/fec.h      | 16 ++++++
 drivers/net/ethernet/freescale/fec_main.c | 69 ++++++++++++++++++-----
 2 files changed, 70 insertions(+), 15 deletions(-)

diff --git a/drivers/net/ethernet/freescale/fec.h b/drivers/net/ethernet/freescale/fec.h
index 5b4da21d8d..5c737634a7 100644
--- a/drivers/net/ethernet/freescale/fec.h
+++ b/drivers/net/ethernet/freescale/fec.h
@@ -527,10 +527,26 @@ struct fec_enet_priv_txrx_info {
 	struct  sk_buff *skb;
 };
 
+#ifdef CONFIG_FEC_OOB
+struct fec_enet_priv_tx_q;
+
+struct fec_inband_work {
+	dma_addr_t addr;
+	struct device *dev;
+	unsigned int size;
+};
+#endif
+
 struct fec_enet_priv_tx_q {
 	struct bufdesc_prop bd;
 	unsigned char *tx_bounce[TX_RING_SIZE];
 	struct  sk_buff *tx_skbuff[TX_RING_SIZE];
+#ifdef CONFIG_FEC_OOB
+	struct fec_inband_work inband_flush[TX_RING_SIZE];
+	struct irq_work inband_irq_work;
+	int next_to_defer;
+	int next_to_flush;
+#endif
 
 	unsigned short tx_stop_threshold;
 	unsigned short tx_wake_threshold;
diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index 7c3e2eb950..27a8109ba3 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -551,12 +551,36 @@ static dma_addr_t get_dma_mapping(struct sk_buff *skb,
 	return addr;
 }
 
-static void release_dma_mapping(struct sk_buff *skb,
-				struct device *dev, dma_addr_t addr, size_t size,
-				enum dma_data_direction dir)
+#ifdef CONFIG_FEC_OOB
+
+static void release_inband_work(struct irq_work *irq_work)
+{
+	struct fec_enet_priv_tx_q *txq = container_of(irq_work, struct fec_enet_priv_tx_q, inband_irq_work);
+
+	while (txq->next_to_flush != READ_ONCE(txq->next_to_defer)) {
+		int ntf = txq->next_to_flush;
+		struct fec_inband_work *fl = txq->inband_flush + ntf;
+		dma_unmap_single(fl->dev, fl->addr, fl->size, DMA_TO_DEVICE);
+		txq->next_to_flush = (ntf + 1) % TX_RING_SIZE;
+	}
+}
+
+static void release_tx_mapping(struct sk_buff *skb,
+			struct fec_enet_priv_tx_q *txq,
+			struct device *dev, dma_addr_t addr, size_t size)
 {
 	if (!fec_net_oob() || !skb || !skb_is_oob_managed(skb)) {
-		dma_unmap_single(dev, addr, size, dir);
+		if (fec_running_oob()) {
+			int ntd = txq->next_to_defer;
+			struct fec_inband_work *fl = txq->inband_flush + ntd;
+			fl->addr = addr;
+			fl->size = size;
+			fl->dev = dev;
+			WRITE_ONCE(txq->next_to_defer, (ntd + 1) % TX_RING_SIZE);
+			irq_work_queue(&txq->inband_irq_work);
+		} else {
+			dma_unmap_single(dev, addr, size, DMA_TO_DEVICE);
+		}
 	} else {
 		/*
 		 * An oob-managed storage should not be unmapped, this
@@ -564,10 +588,21 @@ static void release_dma_mapping(struct sk_buff *skb,
 		 * it belongs to. We only need to synchronize the CPU
 		 * caches for the specified I/O direction.
 		 */
-		dma_sync_single_for_cpu(dev, addr, size, dir);
+		dma_sync_single_for_cpu(dev, addr, size, DMA_TO_DEVICE);
 	}
 }
 
+#else	/* !CONFIG_FEC_OOB */
+
+static void release_tx_mapping(struct sk_buff *skb,
+			struct fec_enet_priv_tx_q *txq,
+			struct device *dev, dma_addr_t addr, size_t size)
+{
+	dma_sync_single_for_cpu(dev, addr, size, DMA_TO_DEVICE);
+}
+
+#endif	/* !CONFIG_FEC_OOB */
+
 static struct bufdesc *
 fec_enet_txq_submit_frag_skb(struct fec_enet_priv_tx_q *txq,
 			     struct sk_buff *skb,
@@ -651,8 +686,8 @@ fec_enet_txq_submit_frag_skb(struct fec_enet_priv_tx_q *txq,
 	bdp = txq->bd.cur;
 	for (i = 0; i < frag; i++) {
 		bdp = fec_enet_get_nextdesc(bdp, &txq->bd);
-		release_dma_mapping(NULL, &fep->pdev->dev, fec32_to_cpu(bdp->cbd_bufaddr),
-				fec16_to_cpu(bdp->cbd_datlen), DMA_TO_DEVICE);
+		release_tx_mapping(NULL, txq, &fep->pdev->dev, fec32_to_cpu(bdp->cbd_bufaddr),
+				fec16_to_cpu(bdp->cbd_datlen));
 	}
 	return ERR_PTR(-ENOMEM);
 }
@@ -717,8 +752,7 @@ static int fec_enet_txq_submit_skb(struct fec_enet_priv_tx_q *txq,
 	if (nr_frags) {
 		last_bdp = fec_enet_txq_submit_frag_skb(txq, skb, ndev);
 		if (IS_ERR(last_bdp)) {
-			release_dma_mapping(skb, &fep->pdev->dev, addr,
-					buflen, DMA_TO_DEVICE);
+			release_tx_mapping(skb, txq, &fep->pdev->dev, addr, buflen);
 			dev_kfree_skb_any(skb);
 			return NETDEV_TX_OK;
 		}
@@ -1088,11 +1122,11 @@ static void fec_enet_bd_init(struct net_device *dev)
 			bdp->cbd_sc = cpu_to_fec16(0);
 			if (bdp->cbd_bufaddr &&
 			    !IS_TSO_HEADER(txq, fec32_to_cpu(bdp->cbd_bufaddr)))
-				release_dma_mapping(txq->tx_skbuff[i],
+				release_tx_mapping(txq->tx_skbuff[i],
+						txq,
 						&fep->pdev->dev,
 						fec32_to_cpu(bdp->cbd_bufaddr),
-						fec16_to_cpu(bdp->cbd_datlen),
-						DMA_TO_DEVICE);
+						fec16_to_cpu(bdp->cbd_datlen));
 			if (txq->tx_skbuff[i]) {
 				dev_kfree_skb_any(txq->tx_skbuff[i]);
 				txq->tx_skbuff[i] = NULL;
@@ -1551,10 +1585,9 @@ fec_enet_tx_queue(struct net_device *ndev, u16 queue_id)
 		skb = txq->tx_skbuff[index];
 		txq->tx_skbuff[index] = NULL;
 		if (!IS_TSO_HEADER(txq, fec32_to_cpu(bdp->cbd_bufaddr)))
-			release_dma_mapping(skb, &fep->pdev->dev,
+			release_tx_mapping(skb, txq, &fep->pdev->dev,
 					fec32_to_cpu(bdp->cbd_bufaddr),
-					fec16_to_cpu(bdp->cbd_datlen),
-					DMA_TO_DEVICE);
+					fec16_to_cpu(bdp->cbd_datlen));
 		bdp->cbd_bufaddr = cpu_to_fec32(0);
 		if (!skb)
 			goto skb_done;
@@ -3408,6 +3441,12 @@ fec_enet_alloc_txq_buffers(struct net_device *ndev, unsigned int queue)
 		}
 
 		bdp = fec_enet_get_nextdesc(bdp, &txq->bd);
+
+#ifdef CONFIG_FEC_OOB
+		init_irq_work(&txq->inband_irq_work, release_inband_work);
+		txq->next_to_defer = 0;
+		txq->next_to_flush = 0;
+#endif
 	}
 
 	/* Set the last buffer to wrap. */
-- 
2.30.2

